pingcap/tidb-operator
# fix: when tls client is enabled for tidb, br will crash the controller manager

### What problem does this PR solve? 

When `TidbCluster.spec.tidb.tlsClient.enabled` is set to `true`, and apply a `Backup` as follows, `tidb-controller-manager` will panic and crash.

fixes #3524

### What is changed and how does it work?

The root cause is that br.spec.from is nil, we should check and access its members

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)
    - apply `backup-s3.yaml` and check that backup files are at the correct s3 bucket, `tidb-controller-manager` is fine
    ```yaml
    # backup-s3.yaml
    apiVersion: pingcap.com/v1alpha1
    kind: Backup
    metadata:
        name: br16
        namespace: ns0
    spec:
        br:
            cluster: ns0
            clusterNamespace: ns0
        s3:
            provider: aws
            region: ap-northeast-1
            secretName: s3-secret
            bucket: backup
            prefix: br16
    ```

Code changes

- Has Go code change

Side effects

None

Related changes

- Need to cherry-pick to the release branch

### Does this PR introduce a user-facing change?:

```release-note
Fix the issue that syncing `Backup` CR will crash tidb-controller-manager when TLS client is enabled for TiDB
```
